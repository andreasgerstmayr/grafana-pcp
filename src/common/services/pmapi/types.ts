import { DataSourceInstanceSettings } from '@grafana/data';
import { InstanceName, Labels, MetricName, Semantics } from 'common/types/pcp';

export interface PmApiConfig {
    dsInstanceSettings: DataSourceInstanceSettings;
    timeoutMs: number;
}

export type InstanceId = number;

export interface Context {
    context: number;
    labels: Labels;
}

export interface MetricResponse {
    metrics: Metadata[];
}

export interface Metadata {
    name: MetricName;
    indom?: string;
    type: string;
    sem: Semantics;
    series: string;
    units: string;
    labels: Labels;
    'text-help': string;
    'text-oneline': string;
}

export interface IndomResponse {
    instances: Indom[];
    labels: Labels;
}

export interface Indom {
    name: InstanceName;
    instance: InstanceId;
    labels: Labels;
}

export interface FetchResponse {
    timestamp: number;
    values: MetricInstanceValues[];
}

export interface MetricInstanceValues {
    name: MetricName;
    instances: InstanceValue[];
}

export interface InstanceValue {
    instance: InstanceId | null;
    value: number | string;
}

export interface StoreResponse {
    success: boolean;
}

export interface DeriveResponse {
    success: boolean;
}

export interface ChildrenResponse {
    leaf: string[];
    nonleaf: string[];
}

export class MetricNotFoundError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Cannot find metric ${metric}. Please check if the PMDA is enabled.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class NoIndomError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Metric ${metric} has no instance domain.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class MetricSemanticError extends Error {
    constructor(readonly expr: string, message?: string) {
        super(message ?? `Semantic error in '${expr}' definition.`);
        this.expr = expr;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class MetricSyntaxError extends Error {
    constructor(readonly expr: string, message?: string) {
        super(message ?? `Syntax error in '${expr}' definition.`);
        this.expr = expr;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class DuplicateDerivedMetricNameError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Duplicate derived metric name ${metric}`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

export class PermissionError extends Error {
    constructor(readonly metric: string, message?: string) {
        super(message ?? `Insufficient permissions to store metric ${metric}. Please check the PMDA configuration.`);
        this.metric = metric;
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
