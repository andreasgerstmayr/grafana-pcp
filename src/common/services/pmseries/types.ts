import { DataSourceInstanceSettings } from '@grafana/data';
import { InstanceName, Labels, Semantics } from 'common/types/pcp';

export interface PmSeriesApiConfig {
    dsInstanceSettings: DataSourceInstanceSettings;
    timeoutMs: number;
    baseUrl: string;
}

export type SeriesId = string;
export type SeriesInstanceId = string;

export interface SeriesNoRecordResponse {
    success: boolean;
}

export interface SeriesDescQueryParams {
    series: string[]; // list of series identifiers
}
export interface SeriesDescItemResponse {
    series: string;
    source: string;
    pmid: string;
    indom: string;
    semantics: Semantics;
    type: string;
    units: string;
}
export type SeriesDescResponse = SeriesDescItemResponse[];

// this one has no related 'maybe' response type
export interface SeriesPingResponse {
    success: boolean;
}

export interface SeriesQueryQueryParams {
    expr: string; // Query string in [pmseries](https://pcp.io/man/man1/pmseries.1.html) format
}
export interface SeriesQueryItemResponse {
    series: string;
    instance: string;
    timestamp: number;
    value: string;
}
export type SeriesQueryResponse = string[] | SeriesQueryItemResponse[];

export interface SeriesInstancesQueryParams {
    /** Comma-separated list of series identifiers */
    series?: string[];
    /** Glob pattern string to match on all labels */
    match?: string;
}
export interface SeriesInstancesItemResponse {
    series: SeriesId;
    source: string;
    instance: SeriesInstanceId;
    id: number;
    name: InstanceName;
}
export type SeriesInstancesResponse = SeriesInstancesItemResponse[];

export interface SeriesLabelsQueryParams {
    series?: string[]; // Comma-separated list of series identifiers
    match?: string; // Glob pattern string to match on all labels
    name?: string; // Find all known label values for given name
    names?: string[]; // Comma-separated list of label names
}
export interface SeriesLabelsItemResponse {
    series: string;
    labels: Labels;
}
export interface SeriesLabelsLabelValuesItemResponse {
    [key: string]: Array<string | number | boolean>;
}
export type SeriesLabelsResponse = string[] | SeriesLabelsItemResponse[] | SeriesLabelsLabelValuesItemResponse;

export interface SeriesMetricsQueryParams {
    series?: string[];
    match?: string;
}
export interface SeriesMetricsItemResponse {
    series: string;
    name: string;
}
export type SeriesMetricsResponse = string[] | SeriesMetricsItemResponse[];

export interface SeriesValuesQueryParams {
    series?: string[];
    samples?: number;
    interval?: string;
    start?: string;
    finish?: string;
    offset?: string;
    align?: string;
    zone?: string;
}
export interface SeriesValuesItemResponse {
    series: string;
    instance?: string;
    /** unix timestamp in milliseconds */
    timestamp: number;
    value: string;
}
export type SeriesValuesResponse = SeriesValuesItemResponse[];
