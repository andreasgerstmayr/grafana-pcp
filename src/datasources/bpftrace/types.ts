import { Optional } from 'utility-types';
import { MinimalPmapiQuery, PmapiOptions, PmapiQuery } from '../../datasources/lib/pmapi/types';
import { TargetFormat } from '../../datasources/lib/types';
import { Script } from './script';

export interface BPFtraceOptions extends PmapiOptions {}

export interface BPFtraceQuery extends MinimalPmapiQuery {}

export const defaultBPFtraceQuery: BPFtraceQuery & Optional<PmapiQuery, 'url' | 'hostspec'> = {
    refId: 'A',
    expr: '',
    format: TargetFormat.TimeSeries,
    options: {
        rateConversation: true,
        timeUtilizationConversation: true,
    },
};

export interface BPFtraceTargetData {
    script: Script;
}
