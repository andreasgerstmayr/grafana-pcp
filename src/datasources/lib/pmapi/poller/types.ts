import { Context, Indom, InstanceValue, Metadata } from 'common/services/pmapi/types';
import { Labels } from 'common/types/pcp';
import { Dict, RequiredField } from 'common/types/utils';
import { Target, TemplatedPmapiQuery } from '../types';

export interface Metric {
    metadata: Metadata;
    instanceDomain?: {
        instances: Dict<string, Indom>;
        labels: Labels;
    };
    values: InstanceValuesSnapshot[];
}

export interface InstanceValuesSnapshot {
    timestampMs: number;
    values: InstanceValue[];
}

export enum EndpointState {
    /** new entered endpoint, no context available */
    PENDING,
    /** context available */
    CONNECTED,
}

/**
 * single endpoint, identified by url and hostspec
 * each url/hostspec has a different context
 * each url/hostspec can have different metrics (and values)
 */
export interface Endpoint {
    state: EndpointState;
    url: string;
    hostspec: string;
    metrics: Metric[];
    targets: Target[];
    additionalMetricsToPoll: Array<{ name: string; callback: (values: InstanceValue[]) => void }>;
    errors: any[];

    /** context, will be created at next poll */
    context?: Context;
    /** backfilling with redis */
    hasRedis?: boolean;
}

export type EndpointWithCtx = RequiredField<Endpoint, 'context'>;

export interface QueryResult {
    endpoint: EndpointWithCtx;
    query: TemplatedPmapiQuery;
    metrics: Metric[];
}
