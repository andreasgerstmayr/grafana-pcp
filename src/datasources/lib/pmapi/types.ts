import { DataQuery, DataSourceJsonData } from '@grafana/data';
import { RequiredField } from 'common/types/utils';
import { TargetFormat } from '../types';

export interface PmapiOptions extends DataSourceJsonData {
    hostspec?: string;
    retentionTime?: string;
}

export interface PmapiDefaultOptions {
    hostspec: string;
    retentionTime: string;
}

export interface PmapiQuery extends DataQuery {
    expr: string;
    format: TargetFormat;
    legendFormat?: string;

    url?: string;
    hostspec?: string;
}

export type TemplatedPmapiQuery = RequiredField<PmapiQuery, 'url' | 'hostspec'>;

export enum TargetState {
    /** newly entered target or target with error (trying again) */
    PENDING,
    /** metrics exists and metadata available */
    METRICS_AVAILABLE,
    /** fatal error, will not try again */
    ERROR,
}

/**
 * Represents a target of a Grafana panel, which will be polled in the background
 * extends the Query (as provided by Grafana) with additional information:
 * - vector+bpftrace: errors occured while polling in the background, lastActive, ...
 * - vector: isDerivedMetric
 * - bpftrace: script
 * is persisted during multiple queries (key = targetId)
 */
export interface Target<T = any> {
    targetId: string;
    state: TargetState;
    query: TemplatedPmapiQuery;
    /** valid PCP metric names (can be a derived metric, e.g. derived_xxx) */
    metricNames: string[];
    errors: any[];
    lastActiveMs: number;
    custom?: T;
}
