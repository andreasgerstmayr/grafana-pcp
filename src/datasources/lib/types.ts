export enum TargetFormat {
    TimeSeries = 'time_series',
    Heatmap = 'heatmap',
    /** vector only */
    MetricsTable = 'metrics_table',
    /** bpftrace only */
    CsvTable = 'csv_table',
    /** bpftrace only */
    FlameGraph = 'flamegraph',
}
