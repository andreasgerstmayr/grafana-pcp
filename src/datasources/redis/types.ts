import { DataQuery, DataSourceJsonData } from '@grafana/data';
import { TargetFormat } from '../../datasources/lib/types';

export interface RedisOptions extends DataSourceJsonData {}

export interface RedisQuery extends DataQuery {
    expr: string;
    format: TargetFormat;
    legendFormat?: string;
}

export const defaultRedisQuery: Partial<RedisQuery> = {
    expr: '',
    format: TargetFormat.TimeSeries,
};
