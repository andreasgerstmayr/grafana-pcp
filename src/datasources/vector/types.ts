import { PmapiOptions, PmapiQuery, TargetFormat } from 'datasources/lib/pmapi/types';

export interface VectorOptions extends PmapiOptions {}

export interface VectorQuery extends PmapiQuery {}

export const defaultVectorQuery: Partial<VectorQuery> = {
    expr: '',
    format: TargetFormat.TimeSeries,
};

export interface VectorTargetData {
    isDerivedMetric: boolean;
}
