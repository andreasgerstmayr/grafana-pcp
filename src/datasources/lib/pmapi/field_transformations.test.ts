import { FieldType, MutableDataFrame } from '@grafana/data';
import { Semantics } from 'common/types/pcp';
import { TargetFormat } from '../types';
import { applyFieldTransformations } from './field_transformations';

describe('field transformations', () => {
    it('should convert time based counters to time utilization', () => {
        const frame = new MutableDataFrame();
        frame.addField({ name: 'Time', type: FieldType.time, values: [2000, 4000, 6000] });
        // first rate:  1,000,000us = 1.0s per sec = CPU was busy 100%
        // second rate:   500,000us = 0.5s per sec = CPU was busy 50%
        frame.addField({ name: 'cgroup.cpu.stat.usage', type: FieldType.number, values: [1000000, 3000000, 4000000] });

        const metadata = {
            name: 'cgroup.cpu.stat.usage',
            series: 'e750369c34fccb1f9c62554dcb9051ff5749b529',
            pmid: '3.67.2',
            indom: '3.16',
            type: 'u64',
            sem: Semantics.Counter,
            units: 'microsec',
            labels: {
                agent: 'proc',
                domainname: 'localdomain',
                hostname: 'agerstmayr-thinkpad',
                machineid: '6dabb302d60b402dabcc13dc4fd0fab8',
            },
            'text-oneline': 'CPU time consumed by processes in each cgroup',
            'text-help': 'CPU time consumed by processes in each cgroup',
        };
        applyFieldTransformations(TargetFormat.TimeSeries, metadata, frame);
        expect(frame).toMatchSnapshot();
    });
});
