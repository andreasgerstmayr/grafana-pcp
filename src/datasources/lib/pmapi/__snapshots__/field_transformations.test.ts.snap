// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field transformations should convert time based counters to time utilization 1`] = `
Object {
  "fields": Array [
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "Time",
      "type": "time",
      "values": Array [
        2000,
        4000,
        6000,
      ],
    },
    Object {
      "config": Object {
        "unit": "percentunit",
      },
      "labels": undefined,
      "name": "cgroup.cpu.stat.usage",
      "type": "number",
      "values": Array [
        0,
        1,
        0.5,
      ],
    },
  ],
  "meta": undefined,
  "name": undefined,
  "refId": undefined,
}
`;
